[Arcilator] Print full values rather than truncating them.

SimonEbner opened the pull request:
Implemented by slicing the value into size_t chunks and passing them as variadic args to printf.

Example:
```
    // CHECK: v8 = fde6741e3a44d997e80d393ae643ee0fd9
    %v8 = arith.constant 0xfde6741e3a44d997e80d393ae643ee0fd9 : i137
    arc.sim.emit "v8", %v8 : i137
```

fzi-hielscher commented on lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp:
```diff
@@ -537,22 +537,59 @@ struct SimEmitValueOpLowering
+    Value value = adaptor.getValue();
+    int sizeTBits = sizeOfSizeT.getFixedValue();
+    auto sizeTType = IntegerType::get(getContext(), sizeTBits);
+    Value shiftValue = arith::ConstantOp::create(
```
Nit: If there is a direct equivalent for the operation in the LLVM dialect I wouldn't take the indirection through `arith`.

SimonEbner commented on test/arcilator/emit-values.mlir:
```diff
@@ -0,0 +1,40 @@
+// RUN: arcilator --run %s | FileCheck %s
+module attributes { dlti.dl_spec = #dlti.dl_spec<
+  #dlti.dl_entry<index, 16>
+> } {
```
i was wondering the same. Updated to use %llx

SimonEbner commented on test/arcilator/emit-values.mlir:
```diff
@@ -0,0 +1,40 @@
+    %v3 = arith.constant 0x0123456789ABCDEF : i64
+    arc.sim.emit "v3", %v3 : i64
+
+    // CHECK: v4 = 0000000000001
```
Makes sense. I was trying not to modify the default behaviour. Updated now to always show all leading zeros.

SimonEbner commented:
I see that by printing leading 0, some of the integration tests fail. Please advice what's the best solution. We could also use a custom printf function that trims all leading zeros correctly (also for >64bit values) which we link into the jit.

fzi-hielscher reviewed (commented):
We'll definitely have to add custom print/formatting functions as part of a runtime library in the future. I had a go at that last year and learned the hard way that, to make it work consistently across platforms, there are a bunch of pitfalls regarding ABIs, libc implementations and linkage. It's doable, but not as straightforward as one would hope. 

Until then simply adapting the expected output of the integration tests to the new format should be fine. You could use a regex to ignore leading zeroes or just explicitly write them in.

SimonEbner commented:
Updated to add either a single leading zero or a 0* regex to the integration tests. I removed the size_t part too and changed slightly the implementation to distinguish between first and other chunks rather than only remainingBits.

OOC, regarding your point on ABI compatible prints: Wouldn't it be possible to create a vararg custom print that takes in int64 values, skips all leading ones that are zero - and starts printing from the first non-null value and rest? Probably not the most efficient but wouldn't that work irrespective of ABI ?

fzi-hielscher reviewed (approved):
Great, thanks!

> OOC, regarding your point on ABI compatible prints: Wouldn't it be possible to create a vararg custom print that takes in int64 values, skips all leading ones that are zero - and starts printing from the first non-null value and rest? Probably not the most efficient but wouldn't that work irrespective of ABI ?

Yeah, that should work. I'd still be hesitant to use a variadic function. The arcane ABI rules make it hard to interface with anything that isn't C, e.g. rust's `c_variadic` [is still unstable.](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html#c_variadic)  I think the best way to deal with this, which should allow us to use types beyond simple integers, is to allocate a "marshaling buffer", manually arrange the raw argument data within it and then pass a pointer and some sort of type descriptor to the external function.

SimonEbner commented on test/Dialect/Arc/lower-sim.mlir:
```diff
@@ -52,14 +50,17 @@ module attributes { dlti.dl_spec = #dlti.dl_spec<
+    // CHECK: %[[lower:.*]] = llvm.trunc %[[arg0]]
+    // CHECK: %[[shifted:.*]] = llvm.lshr %[[arg0]], %[[shift]]
+    // CHECK: %[[higher:.*]] = llvm.trunc %[[shifted]]
+    // %CHECK: llvm.call @printf(%[[format_str3_ptr]], %[[higher]], %[[lower]])
```
thanks, fixed
